### PR TITLE
Fix repository discovery with ceiling_dirs at current directory

### PR DESCRIPTION
**Pull request by joshtriplett**
git only checks ceiling directories when its search ascends to a parent
directory.  A ceiling directory matching the starting directory will not
prevent git from finding a repository in the starting directory or a
parent directory.  libgit2 handled the former case correctly, but
differed from git in the latter case: given a ceiling directory matching
the starting directory, but no repository at the starting directory,
libgit2 would stop the search at that point rather than finding a
repository in a parent directory.

Test case using git command-line tools:

/tmp$ git init x
Initialized empty Git repository in /tmp/x/.git/
/tmp$ cd x/
/tmp/x$ mkdir subdir
/tmp/x$ cd subdir/
/tmp/x/subdir$ GIT_CEILING_DIRECTORIES=/tmp/x git rev-parse --git-dir
fatal: Not a git repository (or any of the parent directories): .git
/tmp/x/subdir$ GIT_CEILING_DIRECTORIES=/tmp/x/subdir git rev-parse --git-dir
/tmp/x/.git

Fix the testsuite to test this case (in one case fixing a test that
depended on the current behavior), and then fix find_repo to handle this
case correctly.

In the process, simplify and document the logic in find_repo():
- Separate the concepts of "currently checking a .git directory" and
  "number of iterations left before going further counts as a search"
  into two separate variables, in_dot_git and min_iterations.
- Move the logic to handle in_dot_git and append /.git to the top of the
  loop.
- Only search ceiling_dirs and find ceiling_offset after running out of
  min_iterations; since ceiling_offset only tracks the longest matching
  ceiling directory, if ceiling_dirs contained both the current
  directory and a parent directory, this change makes find_repo stop the
  search at the parent directory.

### CODE REVIEW

**Comment by joshtriplett** [comment]
I ran into this issue while writing and testing test cases for #3711.  I wrote some test cases for this corner case in `$GIT_CEILING_DIRECTORIES`, found that they failed, and then found I could reproduce the issue using `git_repository_open_ext` or `git_repository_discover` without using any of the new code.

The new behavior, and the new tests, match the way the git command-line tools handle `$GIT_CEILING_DIRECTORIES`.

**Comment by joshtriplett** [comment]
Any thoughts on this bug? Do the test cases seem reasonable, as expressions of git's own behavior? Does the rewritten logic seem like an improvement in clarity?

File: tests/repo/discover.c [hunk 1]
```diff
@@ -118,12 +118,22 @@ void test_repo_discover__0(void)
 	cl_git_fail(git_repository_discover(&found_path, ALTERNATE_MALFORMED_FOLDER3, 0, ceiling_dirs));
 	cl_assert_equal_i(GIT_ENOTFOUND, git_repository_discover(&found_path, ALTERNATE_NOT_FOUND_FOLDER, 0, ceiling_dirs));
 
+	append_ceiling_dir(&ceiling_dirs_buf, SUB_REPOSITORY_FOLDER_SUB);
+	ceiling_dirs = git_buf_cstr(&ceiling_dirs_buf);
+
+	//this must pass as ceiling_directories cannot prevent the current
+	//working directory to be checked
```
Comment (carlosmn) [inline tests/repo/discover.c]: We avoid C++-style comments (it seems this file has at least one instance of this already, but let's not let it get out of control)

File: src/repository.c [hunk 2]
```diff
@@ -359,21 +359,36 @@ static int find_repo(
...
 
 	git_buf_free(repo_path);
 
 	if ((error = git_path_prettify(&path, start_path, NULL)) < 0)
 		return error;
 
-	ceiling_offset = find_ceiling_dir_offset(path.ptr, ceiling_dirs);
+	/* in_dot_git toggles each loop:
+	 * /a/b/c/.git, /a/b/c, /a/b/.git, /a/b, /a/.git, /a
+	 * With GIT_REPOSITORY_OPEN_BARE, we assume we started with /a/b/c.git
```
Comment (carlosmn) [inline src/repository.c]: Do we want to assume this? `_OPEN_BARE` is about avoiding loading things like workdir and config. Anything else should be the same.

File: tests/repo/discover.c (hunk 1 above)
Comment (joshtriplett) [inline tests/repo/discover.c]: No problem, I can change all of the new comments to `/* C style */`.

File: src/repository.c (hunk 2 above)
Comment (joshtriplett) [inline src/repository.c]: That's the defined existing behavior of `_OPEN_BARE`; it already causes the search to not append `/.git`.  I just documented that behavior.

**Comment by joshtriplett** [comment]
Updated to use exclusively `/* C-style comments */`.

**Comment by joshtriplett** [comment]
Following up on this; any issues with the new version of the pull request?

**Comment by joshtriplett** [comment]
>  I think that this is overall a good change, but I'm finding the find_repo function a bit hard to process. This was true before your change, but since we're refactoring it, I wonder if we can clean it up a bit to reduce some of the cognitive load it puts on us.

:+1: It took me quite a while to work through its logic myself, so that sounds good to me.

> It seems like the logic here is correct, but cleaning this up a bit might make it easier to review, if that's something you're willing to do. Thanks!

Sure, I'd be happy to do so.  To some extent, I'd tried to reduce the amount of refactoring to the minimal amount needed to make the logic clearer, because I didn't want to make this more difficult to review.

If you don't mind, I'd like to make this a two-patch series, with the first patch fixing this logic bug and doing some minimal refactoring, and the second doing some more aggressive refactoring.  Does that sound reasonable?

**Comment by ethomson** [comment]
Awesome, thank you.

> If you don't mind, I'd like to make this a two-patch series, with the first patch fixing this logic bug and doing some minimal refactoring, and the second doing some more aggressive refactoring. Does that sound reasonable?

Fine with me!  Thanks for tackling the refactoring.  😀

**Comment by ethomson** [comment]
Included with #3711
